### PR TITLE
Tensor Parallelism training with `neuronx_distributed`

### PR DESCRIPTION
**Pull request by michaelbenayoun**
As per title.

### CODE REVIEW

File: optimum/neuron/accelerate/optimizer.py [hunk 1]
```diff
@@ -40,6 +43,12 @@ def step(self, closure=None):
                 xm.optimizer_step(self.optimizer, optimizer_args=optimizer_args)
             elif self.accelerator_state.distributed_type is NeuronDistributedType.XLA_FSDP:
                 self.optimizer.step(closure)
+            elif self.accelerator_state.distributed_type is NeuronDistributedType.TENSOR_PARALLELISM:
+                xm.optimizer_step(self.optimizer, groups=parallel_state.get_data_parallel_group(as_list=True))
+                # xm.reduce_gradients(self.optimizer, groups=parallel_state.get_data_parallel_group(as_list=True))
+                # parameters = [p for group in self.optimizer.param_groups for p in group["params"]]
+                # parallel_layers.clip_grad_norm(parameters, 1.0, norm_type=2)
+                # self.optimizer.step()
```
Comment (JingyaHuang) [inline optimum/neuron/accelerate/optimizer.py:51]: Do you intend to leave those comments?

File: optimum/neuron/distributed/base.py [hunk 2]
```diff
@@ -78,7 +83,9 @@ def saved_model_in_temporary_directory(cls, model: "PreTrainedModel"):
             tmpdir.cleanup()
 
     @abstractclassmethod
-    def parallelize(cls, model: "PreTrainedModel") -> "PreTrainedModel":
+    def parallelize(
+        cls, model: "PreTrainedModel", return_original_to_transformed_mapping: bool = False
+    ) -> Union["PreTrainedModel", Union["PreTrainedModel", Dict[int, "torch.nn.Parameter"]]]:
```
Comment (JingyaHuang) [inline optimum/neuron/distributed/base.py:88]:
```suggestion
    ) -> Union["PreTrainedModel", Dict[int, "torch.nn.Parameter"]]:
```
?

File: optimum/neuron/utils/cache_utils.py [hunk 3]
```diff
@@ -208,6 +211,12 @@ def get_neuron_cache_path() -> Optional[Path]:
         else:
             path = Path("/var/tmp")
 
+        # TODO: is that correct?
+        # neuronxcc_version = get_neuronxcc_version()
+        # if version.parse(neuronxcc_version) < version.parse(_NEW_CACHE_NAMING_CONVENTION_NEURONXCC_VERSION):
```
Comment (JingyaHuang) [inline optimum/neuron/utils/cache_utils.py:216]: If neuronxcc_version < compiled version, it's not going to work. The inversed is compatible according to Annapurna but I seems encountered issue with backward compatibility either.

**Review by JingyaHuang — APPROVED** [review]
That's huge, congrats on the PR, and thanks for working on it.

Left some minor nits, feel free to merge it whenever you like to.

File: optimum/neuron/accelerate/optimizer.py (hunk 1 above)
Comment (michaelbenayoun) [reply optimum/neuron/accelerate/optimizer.py:51]: Not yet, I will later but for now it might come in handy.

File: optimum/neuron/distributed/base.py (hunk 2 above)
Comment (michaelbenayoun) [reply optimum/neuron/distributed/base.py]: It was an error thanks!

File: optimum/neuron/accelerate/optimizer.py (hunk 1 above)
Comment (michaelbenayoun) [reply optimum/neuron/accelerate/optimizer.py:51]: I tried with it and it still failed. Also we do not do any gradient accumulation in the tests so I am not sure?

File: optimum/neuron/distributed/utils.py [hunk 4]
```diff
@@ -23,19 +23,58 @@
...
+        embedding_layer.embedding_dim,
+        dtype=embedding_layer.weight.dtype,
+    )
+
+    tp_rank = get_tensor_model_parallel_rank()
+    row_size, _ = parallel_embedding_layer.weight.shape
+
+    is_tied = False
+    if lm_head_layer is not None:
+        is_tied = id(embedding_layer.weight.data) == id(lm_head_layer.weight.data)
```
Comment (aws-rhsoln) [inline optimum/neuron/distributed/utils.py:47]: If this is xla_tensor, then the data pointer may not be same. This would mean the weight has to be a CPU tensor first

File: optimum/neuron/accelerate/optimizer.py (hunk 1 above)
Comment (aws-rhsoln) [reply optimum/neuron/accelerate/optimizer.py:51]: grad clipping is unrelated to grad_accumulation. What is the grad_clip value set by default for the non-tp case?
